### DISCUSSION TITLE
Versionamento das Mensagens QR-Code

### DISCUSSION THREAD

**Discussion by celobento** [bacen/pix-api#322] [Ideas] [answered]
Oi Pessoal, boa tarde!

Recentemente foi lançada uma nova versão da API-Pix (2.2.1) e tínhamos feito boa parte da aplicação se baseando na (2.2.0) gerando umas dúvidas na equipe!

1) A partir de quando passa a valer o novo schema da mensagem 2.2.1 ?
2) Quando eu receber mensagens no schema antigo, devo rejeitar?
3) Eu ajustando meu sistema, outros PSPs estarão preparados para aceitar o novo schema?
4) Como consigo identificar qual versão da mensagem estou recebendo, já que não vem na mensagem a versão do schema?

Estas informações são muito importantes, pois dá previsibilidade na evolução da solução.

Trabalhei na equipe NFe da Sefaz e uma informação importante que vinha no XML é a versão da mensagem, além da previsibilidade que tínhamos na mudança de schema.

O BACEN já pensou nestes pontos?

**Comment by ninrod** [comment, marked as answer]
> A partir de quando passa a valer o novo schema da mensagem 2.2.1 ?

A 2.2.X começa a valer a partir de 15 de março de 2021. Diria que é uma boa data guiar o desenvolvimento. 

> Quando eu receber mensagens no schema antigo, devo rejeitar?

A PACS.008 não mudou. Na verdade, nada mudou da 2.2.0 -> 2.2.1. Apenas correções pontuais. 

> Eu ajustando meu sistema, outros PSPs estarão preparados para aceitar o novo schema?

Não existe "novo schema". O schema é para ser compatível. 

> Como consigo identificar qual versão da mensagem estou recebendo, já que não vem na mensagem a versão do schema?

como pagador, você tem que procurar o fragmento `/v2` na URL no local adequado, em se tratando de QR Dinâmico. Se é `/v2`, você sabe que está na 2.2.X a partir de 15 de março. 

Você está mencionando "mensagens", estou entendendo que você está se referindo às PACS. Nada mudou. 

edit: Ressalta-se que não há absolutamente nenhuma obrigação para o PSP Recebedor em termso de oferecer a API Pix. A obrigatoriedade, em se tratando de 15 de março de 2021, é apenas para os PSPs pagadores. Os PSPs pagadores devem estar preparados para entender os endpoints  públicos organizados sob a tag "CobPayload" na versão 2.2.X

**Reply by celobento** [reply]
> P: Quando eu receber mensagens no schema antigo, devo rejeitar?
> R: A PACS.008 não mudou. Na verdade, nada mudou da 2.2.0 -> 2.2.1. Apenas correções pontuais.

O Schema que estou falando é o schema do JSON transitado na API-PIX 

> P: Eu ajustando meu sistema, outros PSPs estarão preparados para aceitar o novo schema?
> R: Não existe "novo schema". O schema é para ser compatível.

Quando falo de novo schema, me refiro a inclusão/exclusão ou obrigatoriedades de campos no schema
![image](https://user-images.githubusercontent.com/4404344/108246485-42527080-7130-11eb-89b1-64b8f0f4e2bb.png)

> P: Como consigo identificar qual versão da mensagem estou recebendo, já que não vem na mensagem a versão do schema?
> R: como pagador, você tem que procurar o fragmento /v2 na URL no local adequado, em se tratando de QR Dinâmico. Se é /v2, você sabe que está na 2.2.X a partir de 15 de março.

Filipe, seria bom termos um meio pra identificar a versão exata a qual se refere a mensagem (json da API-PIX) transitada no QR-CODE. Com certeza o layout da mensagem irá evoluir e o uso de um campo chamado versão facilitaria e tornaria suave a mudança/transição entre as mesmas.

> Você está mencionando "mensagens", estou entendendo que você está se referindo às PACS. Nada mudou.

Me refiro às mensagens em formato JSON usadas para a API-PIX. Uma boa opção é (https://json-schema.org/) para elaborar os schemas.

> edit: Ressalta-se que não há absolutamente nenhuma obrigação para o PSP Recebedor em termso de oferecer a API Pix. A obrigatoriedade, em se tratando de 15 de março de 2021, é apenas para os PSPs pagadores. Os PSPs pagadores devem estar preparados para entender os endpoints públicos organizados sob a tag "CobPayload" na versão 2.2.X

Até para o pagador, validar a mensagem que está sendo recebida é um meio de segurança e a identificação da versão exata da mensagem facilita neste processo e otimiza o tempo de processamento. Vou dar um exemplo:
- 15/03/2021: Passa a valer a versão 2.2.1;
- 10/04/2021: BC lança um novo schema 2.2.2 da API-PIX com novos campos e exclusão de outros e que passará a valer a partir de 15/05/2021 em paralelo com a 2.2.1 até 15/06/2021;
- 15/06/2021: Passa a valer apenas a versão 2.2.2 em produção;

Isso dá uma interoperabilidade entre sistemas e teremos como verificar a versão exata da mensagem que estamos trafegando.

Desde já agradeço a atenção!

**Reply by ninrod** [reply]
@celobento, entendo, acho interessante a sugestão. Nesse caso vale abrir uma issue mesmo lá no repo.

**Reply by celobento** [reply]
Oi Filipe!

Ok, vou registrar uma issue.

Mais uma vez, obrigado pela atenção!
